File: packages/core/src/voyage.ts
import { loadEnv } from "./env.js";
import { safeFetchJson } from "./http.js";

const MODEL = "voyage-3-lite";

type VoyageResp = { data: Array<{ embedding: number[] }> };

export async function embed(texts: string[]): Promise<number[][]> {
  const env = loadEnv();
  if (!env.VOYAGE_API_KEY) {
    // No-op fallback: zero vectors keep the pipeline running for dev/tests.
    return texts.map(() => Array(256).fill(0));
  }
  const res = await safeFetchJson<VoyageResp>("https://api.voyageai.com/v1/embeddings", {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${env.VOYAGE_API_KEY}`,
    },
    body: JSON.stringify({ model: MODEL, input: texts, output_dimension: 256, output_dtype: "float" }),
    timeoutMs: 20_000,
  });
  return res.data.map((d) => d.embedding);
}

export function cosine(a: number[], b: number[]): number {
  let dot = 0;
  let na = 0;
  let nb = 0;
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const ai = a[i] ?? 0;
    const bi = b[i] ?? 0;
    dot += ai * bi;
    na += ai * ai;
    nb += bi * bi;
  }
  if (na === 0 || nb === 0) return 0;
  return dot / Math.sqrt(na * nb);
}

File: core/voyage.py
import math
from .env import load_env
from .http import safe_fetch_json

_MODEL = "voyage-3-lite"


def embed(texts: list[str]) -> list[list[float]]:
    env = load_env()
    if not env.VOYAGE_API_KEY:
        # Zero-vector fallback for dev/test environments without Voyage.
        return [[0.0] * 256 for _ in texts]

    res = safe_fetch_json(
        "https://api.voyageai.com/v1/embeddings",
        method="POST",
        headers={"authorization": f"Bearer {env.VOYAGE_API_KEY}"},
        json_body={
            "model": _MODEL,
            "input": texts,
            "output_dimension": 256,
            "output_dtype": "float",
        },
        timeout_ms=20_000,
    )
    return [d["embedding"] for d in res["data"]]


def cosine(a: list[float], b: list[float]) -> float:
    dot = na = nb = 0.0
    for ai, bi in zip(a, b):
        dot += ai * bi
        na += ai * ai
        nb += bi * bi
    if na == 0 or nb == 0:
        return 0.0
    return dot / math.sqrt(na * nb)
